src/coreclr/src/jit/morph.cpp

```diff
@@ -7279,41 +7287,30 @@ GenTree* Compiler::fgMorphPotentialTailCall(GenTreeCall* call)
         compCurBB->bbJumpKind = BBJ_THROW;
     }
 
-    // For non-void calls, we return a place holder which will be
-    // used by the parent GT_RETURN node of this call.
-
-    GenTree* result = call;
-    if (origCallType != TYP_VOID && info.compRetType != TYP_VOID)
+    if (isRootReplaced)
     {
-        var_types nodeTy = origCallType;
-#ifdef FEATURE_HFA
-        // Return a dummy node, as the return is already removed.
-        if (nodeTy == TYP_STRUCT)
-        {
-            // This is a HFA, use float 0.
-            nodeTy = TYP_FLOAT;
-        }
-#elif defined(UNIX_AMD64_ABI)
-        // Return a dummy node, as the return is already removed.
-        if (varTypeIsStruct(nodeTy))
+        // We have replaced the root node of this stmt and deleted the rest,
+        // but we still have the deleted, dead nodes on the `fgMorph*` stack
+        // if the root node was an `ASG`, `RET` or `CAST`.
+        // Return a zero con node to exit morphing of the old trees without asserts
+        // and forbid POST_ORDER morphing doing something wrong with our call.
+        var_types callType;
+        if (varTypeIsStruct(call))
         {
-            // This is a register-returned struct. Return a 0.
-            // The actual return registers are hacked in lower and the register allocator.
-            nodeTy = TYP_INT;
+            CORINFO_CLASS_HANDLE retClsHnd = call->gtRetClsHnd;
+            callType                       = getReturnTypeForStruct(retClsHnd);
         }
-#endif
-#ifdef FEATURE_SIMD
```

Comment:
In this case if `UNIX_AMD64_ABI` was defined we would have already returned `TYP_INT` because `varTypeIsStruct` returns true for `SIMD` types.

It did not really matter, I tried to return `gtNewZeroConNode(TYP_INT)` and it did not hit any asserts on Win_x64, but that was confusing and become worse with struct changes.